File: packages/shared/src/realtime-lifecycle.test.ts
import { describe, expect, it } from "vitest";

import {
  REALTIME_BACKGROUND_GRACE_MS,
  realtimeDropDelayMs,
  shouldHoldRealtime,
} from "./realtime-lifecycle";

/**
 * #289 — the socket a backgrounded phone should not be holding.
 *
 * Vectors shared with the Kotlin and Swift ports. Two platforms disagreeing
 * about when to drop a socket is how one of them ends up holding it forever,
 * and the symptom of that is a name on the battery screen rather than a bug
 * report.
 */
describe("holding the realtime socket", () => {
  const foreground = {
    visibility: "foreground" as const,
    backgroundedForMs: 0,
    callActive: false,
  };

  it("holds it while somebody is looking at the app", () => {
    expect(shouldHoldRealtime(foreground)).toBe(true);
    expect(realtimeDropDelayMs(foreground)).toBeNull();
  });

  it("holds it through a quick app-switch", () => {
    // Photographing a job, checking an address in Maps, answering a text on a
    // personal line. Tearing the socket down and rebuilding it for each of
    // those costs MORE radio than staying up: a fresh connection is a DNS
    // lookup, a TCP handshake, a TLS handshake and a channel join, against one
    // 300-byte heartbeat.
    expect(
      shouldHoldRealtime({
        visibility: "background",
        backgroundedForMs: 5_000,
        callActive: false,
      }),
    ).toBe(true);
  });

  it("drops it once the phone is genuinely in a pocket", () => {
    expect(
      shouldHoldRealtime({
        visibility: "background",
        backgroundedForMs: REALTIME_BACKGROUND_GRACE_MS,
        callActive: false,
      }),
    ).toBe(false);
  });

  it("never drops it under a live call, however long the app is backgrounded", () => {
    // Call state rides realtime — hold, transfer, the far end hanging up. A
    // call is also exactly when the phone is out of the pocket and often
    // plugged in. Dropping the socket here would trade a saving nobody asked
    // for against a call that silently stops updating.
    for (const backgroundedForMs of [0, REALTIME_BACKGROUND_GRACE_MS, 60 * 60_000]) {
      expect(
        shouldHoldRealtime({
          visibility: "background",
          backgroundedForMs,
          callActive: true,
        }),
        `${backgroundedForMs}ms`,
      ).toBe(true);
      expect(
        realtimeDropDelayMs({
          visibility: "background",
          backgroundedForMs,
          callActive: true,
        }),
      ).toBeNull();
    }
  });
});

describe("when to schedule the drop", () => {
  it("counts down the remaining grace", () => {
    expect(
      realtimeDropDelayMs({
        visibility: "background",
        backgroundedForMs: 10_000,
        callActive: false,
      }),
    ).toBe(REALTIME_BACKGROUND_GRACE_MS - 10_000);
  });

  it("says drop it now once the grace is spent", () => {
    expect(
      realtimeDropDelayMs({
        visibility: "background",
        backgroundedForMs: REALTIME_BACKGROUND_GRACE_MS + 1,
        callActive: false,
      }),
    ).toBe(0);
  });

  it("never returns a negative delay", () => {
    // A phone that was backgrounded overnight comes back with a huge elapsed
    // figure. A negative timer is either an immediate fire or a crash,
    // depending on the platform, and only one of those is the intent.
    const delay = realtimeDropDelayMs({
      visibility: "background",
      backgroundedForMs: 24 * 60 * 60_000,
      callActive: false,
    });
    expect(delay).toBe(0);
  });

  it("agrees with shouldHoldRealtime at every boundary", () => {
    // The two answer the same question — "is the socket wanted" and "when does
    // that change" — and a client wires both. If they can disagree, one
    // platform schedules a drop it then refuses to perform, and the socket
    // stays up forever.
    for (const backgroundedForMs of [
      0,
      1,
      REALTIME_BACKGROUND_GRACE_MS - 1,
      REALTIME_BACKGROUND_GRACE_MS,
      REALTIME_BACKGROUND_GRACE_MS + 1,
    ]) {
      const input = {
        visibility: "background" as const,
        backgroundedForMs,
        callActive: false,
      };
      const held = shouldHoldRealtime(input);
      const delay = realtimeDropDelayMs(input);
      expect(delay === 0, `${backgroundedForMs}ms`).toBe(!held);
    }
  });
});

File: apps/android/app/src/test/kotlin/com/loonext/android/core/realtime/RealtimeLifecycleTest.kt
package com.loonext.android.core.realtime

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #289 — the socket a backgrounded phone should not be holding.
 *
 * Vectors shared with packages/shared/src/realtime-lifecycle.test.ts and the
 * Swift port. Two platforms disagreeing about when to drop a socket is how one
 * of them ends up holding it forever, and the symptom of that is a name on the
 * battery screen rather than a bug report.
 */
class RealtimeLifecycleTest {

    @Test
    fun `holds it while somebody is looking at the app`() {
        assertTrue(RealtimeLifecycle.shouldHold(true, 0, false))
        assertNull(RealtimeLifecycle.dropDelayMs(true, 0, false))
    }

    @Test
    fun `holds it through a quick app-switch`() {
        // Photographing a job, checking an address in Maps, answering a text on
        // a personal line. Tearing the socket down and rebuilding it for each
        // of those costs MORE radio than staying up: a fresh connection is a
        // DNS lookup, a TCP handshake, a TLS handshake and a channel join,
        // against one 300-byte heartbeat.
        assertTrue(RealtimeLifecycle.shouldHold(false, 5_000, false))
    }

    @Test
    fun `drops it once the phone is genuinely in a pocket`() {
        assertFalse(
            RealtimeLifecycle.shouldHold(false, RealtimeLifecycle.BACKGROUND_GRACE_MS, false),
        )
    }

    @Test
    fun `never drops it under a live call`() {
        // Call state rides realtime — hold, transfer, the far end hanging up.
        // A call is also exactly when the phone is out of the pocket and often
        // plugged in.
        for (backgrounded in listOf(0L, RealtimeLifecycle.BACKGROUND_GRACE_MS, 3_600_000L)) {
            assertTrue("$backgrounded", RealtimeLifecycle.shouldHold(false, backgrounded, true))
            assertNull(RealtimeLifecycle.dropDelayMs(false, backgrounded, true))
        }
    }

    @Test
    fun `counts down the remaining grace`() {
        assertEquals(
            RealtimeLifecycle.BACKGROUND_GRACE_MS - 10_000,
            RealtimeLifecycle.dropDelayMs(false, 10_000, false),
        )
    }

    @Test
    fun `never returns a negative delay`() {
        // A phone backgrounded overnight comes back with a huge elapsed figure,
        // and a negative delay() throws.
        assertEquals(0L, RealtimeLifecycle.dropDelayMs(false, 86_400_000, false))
    }

    @Test
    fun `agrees with shouldHold at every boundary`() {
        // The two answer the same question — "is the socket wanted" and "when
        // does that change" — and the view model wires both. If they can
        // disagree, the app schedules a drop it then refuses to perform and the
        // socket stays up forever.
        for (backgrounded in listOf(
            0L,
            1L,
            RealtimeLifecycle.BACKGROUND_GRACE_MS - 1,
            RealtimeLifecycle.BACKGROUND_GRACE_MS,
            RealtimeLifecycle.BACKGROUND_GRACE_MS + 1,
        )) {
            val held = RealtimeLifecycle.shouldHold(false, backgrounded, false)
            val delay = RealtimeLifecycle.dropDelayMs(false, backgrounded, false)
            assertEquals("$backgrounded", !held, delay == 0L)
        }
    }

    @Test
    fun `holds the shared grace window`() {
        // Pinned against packages/shared/src/realtime-lifecycle.ts. A phone
        // that waited a different amount of time from the other would be a
        // silent divergence in the one behaviour this issue is about.
        assertEquals(30_000L, RealtimeLifecycle.BACKGROUND_GRACE_MS)
    }
}
